EKF/mag_control.cpp

```diff
@@ -0,0 +1,355 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2019 Estimation and Control Library (ECL). All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name ECL nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file mag_control.cpp
+ * Control functions for ekf magnetic field fusion
+ */
+
+#include "ekf.h"
+#include <mathlib/mathlib.h>
+
+void Ekf::controlMagFusion()
+{
+	// If we are on ground, store the local position and time to use as a reference
+	// Also reset the flight alignment flag so that the mag fields will be re-initialised next time we achieve flight altitude
+	if (!_control_status.flags.in_air) {
+		_last_on_ground_posD = _state.pos(2);
+		_control_status.flags.mag_aligned_in_flight = false;
+		_num_bad_flight_yaw_events = 0;
+	}
+
+	if ((_params.mag_fusion_type >= MAG_FUSE_TYPE_NONE)
+	    || _control_status.flags.mag_fault
+	    || !_control_status.flags.tilt_align) {
+		stopMagFusion();
+		return;
+	}
+
+	if (canRunMagFusion()) {
+		if (_control_status.flags.in_air) {
+			checkHaglYawResetReq();
+			runInAirYawReset();
+			runVelPosReset();
```

Comment:
The `_velpos_reset_request` flag is currently only set in `realignYawGPS` which is only called -as you said- for fixed-wings. However, I don't see why this flag should be reserved for fixed-wings; the logic isn't there yet, but it would make sense to me that a large heading realignment on a multicopter triggers a vel/pos reset. Don't you think so?

Reply:
Independent of platform type it could be conditional on the level of velocity and position innovation test ratios. If the vehicle is hovering or on ground and has small velocities, then a velocity and position reset is probably unnecessary. When doing a reset to GPS there is always a possibility of resetting to a noise spike.